src/material.cpp

```diff
@@ -353,6 +353,35 @@ Material::~Material()
   model::material_map.erase(id_);
 }
 
+Material & Material::clone()
+{
+  std::unique_ptr<Material> mat = std::make_unique<Material>();
+  mat->index_ = model::materials.size();
+  mat->set_id(C_NONE);
+
+  // set all other parameters to whatever the calling Material has
+  mat->name_ = name_;
+  mat->nuclide_ = nuclide_;
+  mat->element_ = element_;
+  mat->ncrystal_mat_ = ncrystal_mat_;
+  mat->atom_density_ = atom_density_;
+  mat->density_ = density_;
+  mat->density_gpcc_ = density_gpcc_;
+  mat->volume_ = volume_;
+  mat->fissionable_ = fissionable_;
+  mat->depletable_ = depletable_;
+  mat->p0_ = p0_;
+  mat->mat_nuclide_index_ = mat_nuclide_index_;
+  mat->thermal_tables_ = thermal_tables_;
+  mat->temperature_ = temperature_;
+
+  if (ttb_)
+    mat->ttb_ = std::make_unique<Bremsstrahlung>(*ttb_);
+
+  model::materials.push_back(std::move(mat));
+  return *model::materials.back();
+}
```

Comment:
This is a rather small nit pick, but I think we should move these to the end of the method, right before we add this material to the global vector. This will keep an invalid entry from appearing in the `model::material_map` in case something goes wrong between these lines and the ones at the end.

It also localizes the lines making changes to data structures in the `model` namespace.

```suggestion
Material & Material::clone()
{
  std::unique_ptr<Material> mat = std::make_unique<Material>();

  // set all other parameters to whatever the calling Material has
  mat->name_ = name_;
  mat->nuclide_ = nuclide_;
  mat->element_ = element_;
  mat->ncrystal_mat_ = ncrystal_mat_;
  mat->atom_density_ = atom_density_;
  mat->density_ = density_;
  mat->density_gpcc_ = density_gpcc_;
  mat->volume_ = volume_;
  mat->fissionable_ = fissionable_;
  mat->depletable_ = depletable_;
  mat->p0_ = p0_;
  mat->mat_nuclide_index_ = mat_nuclide_index_;
  mat->thermal_tables_ = thermal_tables_;
  mat->temperature_ = temperature_;

  if (ttb_)
    mat->ttb_ = std::make_unique<Bremsstrahlung>(*ttb_);

  mat->index_ = model::materials.size();
  mat->set_id(C_NONE);
  model::materials.push_back(std::move(mat));
  return *model::materials.back();
}
```

Reply:
Not sure why the diff came out like that, but it's just moving these couple lines to the end

```
  mat->index_ = model::materials.size();
  mat->set_id(C_NONE);
```

Reply:
Stylistically, I like this - but in practice, I don't think it's possible for "something to go wrong" with the other lines, because in principle, the calling `Material` would have hit some error when it was being initialized (in other words, the values we're taking from the `Material` were already vetted (that is my understanding at least).

But I have no qualms with this, so I'll change it.

Reply:
Based on the data members that are in the `Material` class now, I think it's highly unlikely that something would go wrong, but, in the future, if we were calling copy constructors of other, more complex objects this reorganization may help us out in diagnosing issues. Thanks for being open to it!

Reply:
true true, that could definitely change in the future